### PR TITLE
feat: url assets

### PR DESCRIPTION
**Pull request by vankop**
**What kind of change does this PR introduce?**

feature: support `new URL("./image.png", import.meta.url)` syntax

**Did you add tests for your changes?**

yes

**Does this PR introduce a breaking change?**

no

**What needs to be documented once your changes are merged?**

webpack supports new syntax for assets `new URL("./image.png", import.meta.url)` => should result to 

``` js
new URL(__webpack_public_path__ + "a1b2d3.png", document.baseURI); // target: web-like
new URL(__webpack_public_path__ + "a1b2d3.png", self.location); // target: webworker
new URL(__webpack_public_path__ + "a1b2d3.png", require("url").pathToFileUrl(__filename)); // target: node-like
```

### CODE REVIEW

File: lib/NormalModuleFactory.js [hunk 1]
```diff
@@ -524,11 +527,15 @@ class NormalModuleFactory extends ModuleFactory {
...
 
+				const dependencyType =
+					dependencies.length > 0 ? dependencies[0].category : undefined;
+
 				// resource with scheme
 				if (scheme) {
 					resourceData = {
 						resource: unresolvedResource,
 						data: {},
+						dependencyType,
```
Comment (sokra) [inline lib/NormalModuleFactory.js:538]: This should not be put into the resource data. It's not info about the resource. Just pass `dependencyType` directly to the ruleset

File: lib/config/defaults.js [hunk 2]
```diff
@@ -466,6 +468,14 @@ const applyModuleDefaults = (
 				...wasm
 			});
 		}
+		if (asset) {
+			rules.push({
+				test: /\.js$/,
```
Comment (sokra) [inline lib/config/defaults.js:473]: Doesn't make sense here. new URL should work for all files. This is not the origin, but the target.

File: test/configCases/asset-url/target-web1/test.config.js [hunk 3]
```diff
@@ -0,0 +1,9 @@
+let _URL = require("url").URL;
+
+module.exports = {
+	moduleScope(scope) {
+		scope.URL = function URL(a, b) {
```
Comment (sokra) [inline test/configCases/asset-url/target-web1/test.config.js:5]: This should be exposed by the test template

File: test/configCases/asset-url/target-node2/webpack.config.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+module.exports = {
+	mode: "development",
+	target: "node",
+	devtool: false,
+	output: {
+		assetModuleFilename: "[name][ext]"
+	},
+	module: {
+		rules: [
+			{
```
Comment (sokra) [inline test/configCases/asset-url/target-node2/webpack.config.js:11]: These test cases should not have a rule for that. The default rule should cover that.

File: lib/BaseURIRuntimeModule.js [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+		super("baseURI");
+		this.env = environment;
+	}
+
+	/**
+	 * @returns {string} runtime code
+	 */
+	generate() {
+		switch (this.env) {
+			case "web":
```
[the remark this answers is not in the document]
Comment (vankop) [reply lib/BaseURIRuntimeModule.js]: Not sure that I did what you expected 🤷‍♂️

File: lib/config/defaults.js (hunk 2 above)
Comment (vankop) [reply lib/config/defaults.js:473]: ah yes, thanks for pointing

File: lib/web/JsonpChunkLoadingRuntimeModule.js [hunk 6]
```diff
@@ -98,6 +99,12 @@ class JsonpChunkLoadingRuntimeModule extends RuntimeModule {
 			chunkGraph.getChunkConditionMap(chunk, chunkHasJs)
 		);
 		return Template.asString([
+			withBaseURI
+				? Template.asString([
+						`${RuntimeGlobals.baseURI} = document.baseURI || self.location.href;`
```
Comment (vankop) [inline lib/web/JsonpChunkLoadingRuntimeModule.js:104]: technically this also should work
```suggestion
						`${RuntimeGlobals.baseURI} = document.baseURI || self.location;`
```

Comment (sokra) [reply lib/web/JsonpChunkLoadingRuntimeModule.js:104]: I thought about making in consistent always a string...

**Review by sokra — APPROVED** [review]

**Comment by sokra** [comment]
Thanks

**Comment by webmaster128** [comment]
Amazing feature, thank you! This allows us to use Wasm in npm packages that can be used in Web, Node and Webpack projects.

Did you consider not requiring `./` at the beginning of the relative path? Both

```js
let a = new URL("./wasm/drand.wasm", "file:///Users/me/projects/drand-client/verifying.js");
let b = new URL("wasm/drand.wasm", "file:///Users/me/projects/drand-client/verifying.js");
```

are the same URL but the second one fails with

> Module not found: Error: Can't resolve 'wasm/drand.wasm' in '/Users/simon/opensource/drand/drand-client'

**Comment by vankop** [comment]
@webmaster128 `new URL("wasm/drand.wasm")` is not valid url

**Comment by webmaster128** [comment]
@vankop right, but it is when a base URL is provided in the second parameter. There are such [examples on MDN](https://developer.mozilla.org/en-US/docs/Web/API/URL/URL#Examples).

I'm not super familiar with [the spec](https://url.spec.whatwg.org/#constructors), but it says

> A base URL is necessary if the input is a relative-URL string: 

where [a relative-URL string](https://url.spec.whatwg.org/#relative-url-string) for http/https/file can be a [path-relative-scheme-less-URL string](https://url.spec.whatwg.org/#path-relative-scheme-less-url-string), which does not seem to require a leading dot.

**Pull request merged**